lib/go/test/flow_epoch_test.go

```diff
@@ -1516,3 +1517,203 @@ func TestEpochReset(t *testing.T) {
+		stakingPublicKeys,
+		networkingPublicKeys,
+		ids)
+
+	// Set the approved node list
+	tx := createTxWithTemplateAndAuthorizer(b, templates.GenerateSetApprovedNodesScript(env), idTableAddress)
+
+	approvedNodeIDs := generateCadenceNodeDictionary(ids)
+	err := tx.AddArgument(approvedNodeIDs)
+	require.NoError(t, err)
+
+	signAndSubmit(
+		t, b, tx,
+		[]flow.Address{idTableAddress},
+		[]sdkcrypto.Signer{IDTableSigner},
+		false,
+	)
+
+	// Advance to epoch Setup and make sure that the epoch cannot be ended
+	advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "EPOCHSETUP", false)
+
+	// Register a QC voter
+	tx = createTxWithTemplateAndAuthorizer(b, templates.GenerateEpochRegisterQCVoterScript(env), addresses[0])
+	signAndSubmit(
+		t, b, tx,
+		[]flow.Address{addresses[0]},
+		[]sdkcrypto.Signer{signers[0]},
+		false,
+	)
+
+	tx = createTxWithTemplateAndAuthorizer(b, templates.GenerateEpochRegisterQCVoterScript(env), addresses[5])
+	signAndSubmit(
+		t, b, tx,
+		[]flow.Address{addresses[5]},
+		[]sdkcrypto.Signer{signers[5]},
+		false,
+	)
+
+	// Register a DKG Participant
+	tx = createTxWithTemplateAndAuthorizer(b, templates.GenerateEpochRegisterDKGParticipantScript(env), addresses[1])
+	signAndSubmit(
+		t, b, tx,
+		[]flow.Address{addresses[1]},
+		[]sdkcrypto.Signer{signers[1]},
+		false,
+	)
+
+	// Perform epoch recovery with a new epoch and epoch recover overwriting the current epoch.
+	t.Run("Can recover the epoch and have everything return to normal", func(t *testing.T) {
+		epochTimingConfigResult := executeScriptAndCheck(t, b, templates.GenerateGetEpochTimingConfigScript(env), nil)
+
+		var (
+			startView      uint64 = 100
+			stakingEndView uint64 = 120
+			endView        uint64 = 160
+			targetDuration uint64 = numEpochViews
+			targetEndTime  uint64 = expectedTargetEndTime(epochTimingConfigResult, startEpochCounter+1)
+		)
+
+		collectorClusters := make([]cadence.Value, 3)
+		collectorClusters[0] = cadence.NewArray([]cadence.Value{CadenceString("node_1"), CadenceString("node_2"), CadenceString("node_3")})
+		collectorClusters[1] = cadence.NewArray([]cadence.Value{CadenceString("node_4"), CadenceString("node_5"), CadenceString("node_6")})
+		collectorClusters[2] = cadence.NewArray([]cadence.Value{CadenceString("node_7"), CadenceString("node_8"), CadenceString("node_9")})
+
+		dkgPubKeys := []string{"pubkey_1"}
+		dkgPubKeysCdc := make([]cadence.Value, len(dkgPubKeys))
+		for i, key := range dkgPubKeys {
+			dkgPubKeysCdc[i], _ = cadence.NewString(key)
+		}
+
+		nodeIDs := make([]cadence.Value, len(ids))
+		for i, id := range ids {
+			nodeIDs[i], _ = cadence.NewString(id)
+		}
+
+		args := []cadence.Value{
+			cadence.NewUInt64(startView),
+			cadence.NewUInt64(stakingEndView),
+			cadence.NewUInt64(endView),
+			cadence.NewUInt64(targetDuration),
+			cadence.NewUInt64(targetEndTime),
+			cadence.NewArray(collectorClusters),        // collectorClusters
+			cadence.NewArray(make([]cadence.Value, 0)), // clusterQCVoteData
```

Comment:
Given that this seems to work (3 clusters and 0 QCs), perhaps we should add a sanity check to `recoverEpoch` that `len(clusters) == len(clusterQCs)`

Reply:
https://github.com/onflow/flow-core-contracts/pull/420/commits/a78b7d7ce5e271717b04aaa6f289612b01e4d6ea